### PR TITLE
Notify listeners of async RealmObject even if the result is empty

### PR DESCRIPTION
**Pull request by nhachicha**
Fixes #2218 #2200

### CODE REVIEW

**Comment by nhachicha** [comment]
retest this please

**Comment by nhachicha** [comment]
@realm/java

File: realm/realm-library/src/androidTest/java/io/realm/TypeBasedNotificationsTests.java [hunk 1]
```diff
@@ -540,6 +540,11 @@ public void onChange() {
             public void onChange() {
                 switch (typebasedCommitInvocations.incrementAndGet()) {
                     case 1: {
+                        assertTrue(dog.isLoaded());
+                        assertFalse(dog.isValid());
+                        break;
+                    }
```
Comment (cmelchior) [inline realm/realm-library/src/androidTest/java/io/realm/TypeBasedNotificationsTests.java]: We don't use blocks but `break` in the rest of the codebase.

File: realm/realm-library/src/androidTest/java/io/realm/TypeBasedNotificationsTests.java [hunk 2]
```diff
@@ -619,7 +624,13 @@ public void onChange() {
             @Override
             public void onChange() {
                 switch (typebasedCommitInvocations.incrementAndGet()) {
-                    case 1: {
+                    case 1:  // triggered by COMPLETED_ASYNC_REALM_OBJECT
+                    case 2: {// triggered by the irrelevant commit (not affecting Dog table)
+                        assertTrue(dog.isLoaded());
+                        assertFalse(dog.isValid());
+                        break;
+                    }
```
Comment (cmelchior) [inline realm/realm-library/src/androidTest/java/io/realm/TypeBasedNotificationsTests.java]: Same

File: realm/realm-library/src/androidTest/java/io/realm/TypeBasedNotificationsTests.java [hunk 3]
```diff
@@ -640,7 +651,7 @@ public void run() {
                         });
                         break;
                     }
-                    case 2: {
+                    case 4: {
```
Comment (cmelchior) [inline realm/realm-library/src/androidTest/java/io/realm/TypeBasedNotificationsTests.java]: Same

File: realm/realm-library/src/main/java/io/realm/RealmObject.java [hunk 4]
```diff
@@ -309,11 +309,24 @@ public final void removeChangeListeners() {
...
     void notifyChangeListeners() {
         if (listeners != null && !listeners.isEmpty()) {
-            if (row.getTable() == null) return;
+            boolean notify = false;
 
-            long version = row.getTable().version();
-            if (currentTableVersion != version) {
-                currentTableVersion = version;
+            Table table = row.getTable();
+            if (table == null) {
```
Comment (cmelchior) [inline realm/realm-library/src/main/java/io/realm/RealmObject.java]: Is `notifyListeners` only guaranteed to be called when a async RealmObject query is completed? Couldn't this also be triggered if some unrelated async RealmResults query completed?

Comment (nhachicha) [inline realm/realm-library/src/main/java/io/realm/RealmObject.java]: All invocation scenarios involve an async ops, except `refresh` & local commits that will also trigger the listeners

**Comment by nhachicha** [comment]
squashed

**Comment by cmelchior** [comment]
:+1:

**Pull request merged**